gpii/node_modules/flowManager/src/GpiiRevisionRequester.js

```diff
@@ -0,0 +1,84 @@
+/*!
+GPII Full SHA Revision DataSource
+
+Copyright 2020 OCAD University
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://github.com/GPII/universal/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+var fluid = require("infusion"),
+    gpii = fluid.registerNamespace("gpii");
+
+require("kettle");
+
+fluid.registerNamespace("gpii.flowmanager.revisionRequester");
+
+fluid.defaults("gpii.flowmanager.revisionRequester", {
+    gradeNames: ["fluid.component"],
+
+    // Prepend CBFM host/port distributed down from, e.g., gpii.flowManager.config.untrusted
+    // For development tests where there is no valid /revision endpoint in the
+    // cloud, this is left as null.
+    cloudURL: null,
+    urlTemplate: "%cloudURL/revision",
+
+    // Compute url from CBFM base URL and template
+    revisionGetUrl: {
+        expander: {
+            funcName: "fluid.stringTemplate",
+            args: ["{that}.options.urlTemplate", {
+                cloudURL: "{that}.options.cloudURL"
+            }]
+        }
+    },
+    components: {
+        gpiiRevisionDataSource: {
+            type: "kettle.dataSource.URL",
+            options: {
+                url: "{revisionRequester}.options.revisionGetUrl"
+            }
+        }
+    },
+    invokers: {
+        getRevision: {
+            funcName: "gpii.flowmanager.revisionRequester.getRevision",
+            args: ["{that}"]
+        }
+    }
+});
+
+/**
+ * Retrieve the respository revision's full SHA256 from the cloud.
+ * @param {Component} that - An instance of gpii.flowManager.revisionRequester.
+ * @return {Promise} A promise whose resolved value is eiher the revision or, if
+ * there is an error, an object with an "isError: true" property.  The revision
+ * has the form { "sha256": "86a83d2f93a6f8f954a4fef618ca6aea1399c980" }.
+ */
+gpii.flowmanager.revisionRequester.getRevision = function (that) {
+    var togo = fluid.promise();
+
+    if (that.options.cloudURL !== null) {
+        var revisionPromise = that.gpiiRevisionDataSource.get();
+        revisionPromise.then(function (/*revision*/) {
+            fluid.promise.follow(revisionPromise, togo);
+        }, function (err) {
+            togo.resolve(err);
+        });
+    } else {
+        // If the url to the CBFM is null, assume this is running in a
+        // development testing enviroment where there is no valid /revision
+        // endpoint.
+        togo.resolve({
```

Comment:
After thinking a bit more about it, rather than resolving an error message, it might make more sense to `togo.resolve(null)` as `cloudURL === null` is a intentional and supported case.

In the documentation, please add that setting cloudURL to `null` is the approach to stop LFM fetch SR from the Cloud but using local SR instead.

Reply:
Add to the above, `gpii.flowManager.solutionsRegistry.dataSource.local.loadFromRepository` will look like:
```
if (!revision) {
    togo.resolve(null);
} else {
    fetch SR from Cloud; 
}
```

Reply:
At the risk of derailing myself:  I am and have been uncomfortable using `resolve()` here instead of `reject()`.  After all, it is a failure and normally that means rejecting the promise.  However, the problem with `reject()` is that the sequence would come to a halt, or at least that's my understanding of what happens with `fluid.promise.fireTransformEvent()` -- each transform is executed until one of them is rejected, or until all of them are complete.  In the case of the SRDS, the `solutionsRegistryReady` event is fired by the last step, but only if the whole sequence is successful (?).

Given that, would a reject() be an option if there were a way to include a rejection handler for the whole sequence, that fires the required `solutionsRegistryReady` event?  If that were possible, a single reject() would abort the rest of the sequence, and there is no need for resolving to null (as you suggest), or to a payload with an error and/or status code (as I've written it).  The rejection would invoke this proposed handler to call "SRDS.events.solutionsRegistryReady".  Or is that just not how it works with a transform event sequence?

I'll explore it a bit.

Reply:
My understanding is the same that the rejection here will cause the `fluid.promise.fireTransformEvent()` chain to reject from here. It won't proceed with the rest actions.

I feel using `resolve()` makes sense here because `url === null` is the way-out case we support for telling the system not to fetch from remote but use the local instead.

Reply:
Okay, but FWIW, I took a close look at how `fluid.promise.fireTransformEvent()` works.  The call returns a `promise` that can be thought of as the promise for the whole sequence.  A rejection handler can be added to it that, in this case, fires the `solutionsRegistryReady` event.  Then, a rejection at any point in the load sequence, (1) ends the sequence immediately, (2) fires the "solutionsRegistryReady" event.

Having said that, I didn't use it, and went with a version of your suggestion to test the revision, and use `resolve()`.  I did, however, test the idea and, if you are interested, have a look at this gist: https://gist.github.com/klown/a4f13aa6b2207e82d403c38ffb20ecd1